### PR TITLE
feat: add tracing helper

### PR DESCRIPTION
**Pull request by alnr**

### CODE REVIEW

**Review by aeneasr — COMMENTED** [review]
What an interesting solution! I was not aware that we can solve the overshadowing problem by pointing to the error variable. The only rare cases I can think of where this would not work is when we create a second (inner) error variable:

```
func foo() (err error) {
  if err := baz(); err != nil {
    a, innerErr := bar()
    if innerErr != nil {
      return innerErr
    }
    return err
  }
}
```

But in those cases, we can deal with the result in the code itself by introducing for example a second helper:

```go
func foo() (err error) {
  span, ctx := ...
  defer End(span, &err)
  if err := baz(); err != nil {
    a, innerErr := bar()
    if innerErr != nil {
      return tracerx.TraceError(span, innerErr)
    }
    return err
  }
}
```

Generally speaking, this solution is a 10x improvement to the callback one. I really like it :)

**Review by CaptainStandby — APPROVED** [review]
Neat little trick :+1:

File: otelx/withspan.go [hunk 1]
```diff
@@ -21,21 +21,54 @@ func WithSpan(ctx context.Context, name string, f func(context.Context) error, o
...
+	switch e := recovered.(type) {
+	case error:
+		span.RecordError(e)
+		span.SetStatus(codes.Error, "panic: "+e.Error())
+	case interface{ String() string }:
+		span.SetStatus(codes.Error, "panic: "+e.String())
+	case string:
+		span.SetStatus(codes.Error, "panic: "+e)
+	default:
+		span.SetStatus(codes.Error, "panic")
```
[the remark this answers is not in the document]
Comment (alnr) [reply otelx/withspan.go]: done

**Comment by alnr** [comment]
> What an interesting solution! I was not aware that we can solve the overshadowing problem by pointing to the error variable. The only rare cases I can think of where this would not work is when we create a second (inner) error variable:
> 
> ```
> func foo() (err error) {
>   if err := baz(); err != nil {
>     a, innerErr := bar()
>     if innerErr != nil {
>       return innerErr
>     }
>     return err
>   }
> }
> ```

This works with my solution 😊 I've added a test to show that. Golang will assign the return value (whatever it's name, if it even has one) to the named return value before running deferred calls, so this solution is always correct.


> Generally speaking, this solution is a 10x improvement to the callback one. I really like it :)

Thanks 🥰. `WithSpan` and `End` are complementary in my mind. We should use what's more appropriate.

**Review by CaptainStandby — APPROVED** [review]

**Review by aeneasr — APPROVED** [review]

**Pull request merged**